deps/querystring.lua

```diff
@@ -25,84 +25,89 @@ limitations under the License.
   tags = {"luvit", "url", "codec"}
 ]]
 
-local find = string.find
-local gsub = string.gsub
-local char = string.char
-local byte = string.byte
-local format = string.format
-local match = string.match
-local gmatch = string.gmatch
+
+local function hexToChar(hex)
+  return string.char(tonumber(hex, 16))
+end
+
+local function charToHex(character)
+  return string.format('%%%02X', string.byte(character))
+end
+
 
 local function urldecode(str)
-  str = gsub(str, '+', ' ')
-  str = gsub(str, '%%(%x%x)', function(h)
-    return char(tonumber(h, 16))
-  end)
-  str = gsub(str, '\r\n', '\n')
+  if str then
+    str = string.gsub(str, '+', ' ')
+    str = string.gsub(str, '%%(%x%x)', hexToChar)
+  end
   return str
 end
 
 local function urlencode(str)
   if str then
-    str = gsub(str, '\n', '\r\n')
-    str = gsub(str, '([^%w-_.~])', function(c)
-      return format('%%%02X', byte(c))
-    end)
+    str = string.gsub(str, '[^a-zA-Z0-9*%-%._]', charToHex)
```

Comment:
Why did you remove `~`?

Reply:
According to [the spec](https://url.spec.whatwg.org/#application/x-www-form-urlencoded), 
> The [application/x-www-form-urlencoded percent-encode set](https://url.spec.whatwg.org/#application-x-www-form-urlencoded-percent-encode-set) contains all code points, except the [ASCII alphanumeric](https://infra.spec.whatwg.org/#ascii-alphanumeric), U+002A (*), U+002D (-), U+002E (.), and U+005F (_).

Reply:
[RFC 3986](https://www.rfc-editor.org/rfc/rfc3986#section-2.3) excludes `alphanum`, `-`, `.`, `_`, and `~`.

[Node](https://github.com/nodejs/node/blob/main/lib/querystring.js#L140-L155) excludes more than that.

Why are these all different?

Reply:
Node's probably doing whatever it wants in this regard, it is not incorrect however.

The query component of a URI is defined as:

```
query         = *( pchar / "/" / "?" )
pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"
pct-encoded   = "%" HEXDIG HEXDIG
sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
                 / "*" / "+" / "," / ";" / "="
```

Which means the following don't *have* to be percent-encoded: ALPHA DIGIT - . _ ~ ! $ & ' ( ) * + , ; = / ?